src/App.tsx

```diff
@@ -1,24 +1,196 @@
+  const [loadingTodosId, setLoadingTodosId] = useState<number[]>([0]);
+  const [isDisabled, setIsDisabled] = useState(false);
+  const [isError, setIsError] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorMessage, setErrorMessage] = useState<ErrorTypes>();
+
+  const titleField = useRef<HTMLInputElement>(null);
+
+  const handleChangeTitle = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(e.target.value);
+  };
+
+  const showAndDeleteError = () => {
+    setIsError(true);
+
+    return setTimeout(() => setIsError(false), 3000);
+  };
+
+  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!title.trim()) {
+      setErrorMessage(ErrorTypes.TITLE);
+      setTitle('');
+
+      return showAndDeleteError();
+    }
+
+    setIsDisabled(true);
+    setTempTodo({
+      id: 0,
+      userId: USER_ID,
+      title,
+      completed: false,
+    });
+
+    addTodos({
+      title,
+      userId: USER_ID,
+      completed: false,
+    })
+      .then((res) => {
+        setTodos(prevTodos => [...prevTodos, res]);
+        setIsDisabled(false);
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setErrorMessage(ErrorTypes.ADD);
+        showAndDeleteError();
+        setIsDisabled(false);
+        setTempTodo(null);
+      });
+
+    return setTitle('');
+  };
+
+  const filterTodos = (filterTodosBy: FilterTypes) => {
+    switch (filterTodosBy) {
+      case FilterTypes.ACTIVE:
+        return todos.filter(todo => !todo.completed);
+
+      case FilterTypes.COMPLETED:
+        return todos.filter(todo => todo.completed);
+
+      case FilterTypes.ALL:
+      default:
+        return todos;
+    }
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodosId(prevTodos => [...prevTodos, todoId]);
+    deleteTodos(todoId)
+      .then(() => {
+        setTodos(prevTodos => [...prevTodos
+          .filter(todo => todo.id !== todoId)]);
+        setLoadingTodosId(prevLoadingTodos => [...prevLoadingTodos
+          .filter(Id => Id !== todoId)]);
+      })
+      .catch(() => {
+        setErrorMessage(ErrorTypes.DELETE);
+        showAndDeleteError();
+        setLoadingTodosId(prevLoadingTodos => [...prevLoadingTodos
+          .filter(Id => Id !== todoId)]);
+      });
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage(ErrorTypes.LOAD);
+        showAndDeleteError();
+      });
+
+    return () => {
+      clearTimeout(showAndDeleteError());
+    };
+  }, []);
+
+  useEffect(() => {
+    if (!isDisabled) {
+      return;
+    }
+
+    setTimeout(() => {
+      if (titleField.current) {
+        titleField.current.focus();
+      }
+    }, 400);
+  }, [isDisabled]);
+
+  const visibleTodos = filterTodos(selectedFilter);
+  const todosLeftToFinish = filterTodos(FilterTypes.ACTIVE);
```

Comment:
make sense to use useMemo here, because it's essential to performance

Reply:
done